### PR TITLE
Anway shmdt syscall

### PR DESCRIPTION
**Pull request by Anway-Agte**
## Description
This PR adds test cases and comments for the shmdt and shmat syscalls

### Type of change

- [ ] This change adds documentations and tests.

## How Has This Been Tested?

- Test A - `ut_lind_fs_shmat_syscall`
- Test B - `ut_lind_fs_shmdt_syscall`

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] Any dependent changes have been added to a pull request and/or merged in other modules (native-client, lind-glibc, lind-project)

### CODE REVIEW

File: src/safeposix/syscalls/fs_calls.rs [hunk 1]
```diff
@@ -5083,59 +5083,127 @@ impl Cage {
...
+    /// 
+    /// ### Panics
+    /// 
+    /// Currently there are no scenarios where the function panics
+    /// 
+    /// For more information - refer the documentation at [https://man7.org/linux/man-pages/man3/shmat.3p.html]
     pub fn shmat_syscall(&self, shmid: i32, shmaddr: *mut u8, shmflg: i32) -> i32 {
+        // Get shm table
         let metadata = &SHM_METADATA;
         let prot: i32;
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/fs_calls.rs:5525]: Why are we using the `prot` flag? I see we are assigning either `PROT_READ` or `PROT_WRITE` flags to this, but couldn't understand it's usecase.

File: src/safeposix/syscalls/fs_calls.rs [hunk 2]
```diff
@@ -5083,59 +5083,127 @@ impl Cage {
...
+    /// 
+    /// This function panics if creating an inode fails
+    /// 
+    /// For more information please refer - [https://man7.org/linux/man-pages/man3/shmdt.3p.html]
     pub fn shmdt_syscall(&self, shmaddr: *mut u8) -> i32 {
         let metadata = &SHM_METADATA;
         let mut rm = false;
+        // Acquire the lock of the reverse memory mappings of the Cage object
         let mut rev_shm = self.rev_shm.lock();
+        // This function returns the index where the pair of (shmaddr, shmid) is store within the vector
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/fs_calls.rs]: *stored

File: src/safeposix/syscalls/fs_calls.rs [hunk 3]
```diff
@@ -5083,59 +5083,127 @@ impl Cage {
...
         let mut rm = false;
+        // Acquire the lock of the reverse memory mappings of the Cage object
         let mut rev_shm = self.rev_shm.lock();
+        // This function returns the index where the pair of (shmaddr, shmid) is store within the vector 
+        // That holds these mappings
         let rev_shm_index = Self::rev_shm_find_index_by_addr(&rev_shm, shmaddr as u32);
 
+        // Check if the index is valid
         if let Some(index) = rev_shm_index {
+            // Get the second element of the (shmaddr,shmid) pair which gives us the id of the memory segment
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/fs_calls.rs]: Please run `cargo fmt` for formatting.

File: src/safeposix/syscalls/fs_calls.rs [hunk 4]
```diff
@@ -5083,59 +5083,127 @@ impl Cage {
...
             let shmid = rev_shm[index].1;
+            // Get the memory segment from the shmtable which corresponds to the shmid extracted above
             match metadata.shmtable.entry(shmid) {
+                // If the memory segment is occupied
```
Comment (namanlalitnyu) [inline src/safeposix/syscalls/fs_calls.rs:5623]: Can you explain, what are we doing in the below section, when the memory segment is occupied?

File: src/tests/fs_tests.rs [hunk 5]
```diff
@@ -6,9 +6,11 @@ pub mod fs_tests {
     use crate::interface;
     use crate::safeposix::syscalls::fs_calls::*;
     use crate::safeposix::{cage::*, dispatcher::*, filesystem};
-    use libc::c_void;
+    use libc::{c_void, EINVAL};
+    use std::borrow::{Borrow, BorrowMut};
```
Comment (namanlalitnyu) [inline src/tests/fs_tests.rs]: Are we using these imports including the "Duration" and "Instant" ones in the tests, currently I don't see their references.

**Review by namanlalitnyu — CHANGES_REQUESTED** [review]
couple of minor comments.

File: src/tests/fs_tests.rs (hunk 5 above)
Comment (Anway-Agte) [reply src/tests/fs_tests.rs]: Thanks for pointing that out!

File: src/safeposix/syscalls/fs_calls.rs (hunk 1 above)
Comment (Anway-Agte) [reply src/safeposix/syscalls/fs_calls.rs:5525]: Yeah seems that is not being used anywhere

Comment (namanlalitnyu) [reply src/safeposix/syscalls/fs_calls.rs:5525]: Its being passed to the `segment.map_shm` function.

**Review by namanlalitnyu — APPROVED** [review]
LGTM!

File: src/tests/fs_tests.rs [hunk 6]
```diff
@@ -4038,31 +4039,99 @@ pub mod fs_tests {
...
+        assert_eq!(
+            cage.shmat_syscall(shmid + 10, 0xfffff000 as *mut u8, 0),
+            -(Errno::EINVAL as i32)
+        );
+
+        lindrustfinalize();
+    }
+
+    #[test]
+    pub fn ut_lind_fs_smhdt_syscall() {
```
Comment (pranav-bhatt) [inline src/tests/fs_tests.rs]: typo

**Review by pranav-bhatt — CHANGES_REQUESTED** [review]

**Review by pranav-bhatt — APPROVED** [review]

**Review by yashaswi2000 — APPROVED** [review]
LGTM

**Comment by rennergade** [comment]
This has some merge conflicts but once those are resolved this is good to go.

**Pull request merged**